scripts/image_builder.py

```diff
@@ -0,0 +1,797 @@
+            file.write(f"{self.magic}\n")
+        for cmd in self.contents:
+            prog_offs, prog_spec = cmd.emit(file, enc, prog_offs, self.is_relative)
+            if prog_spec:
+                progs.append(prog_spec)
+
+        file.write("\0")
+
+        if file.tell() > self.size:
+            raise ValueError(f"Generated user script is too large (allocated size: {self.size} < actual size {file.tell()})")
+
+        return progs
+
+
+def render_val(tpl: Any, **kwargs) -> Any:  # mostly str | List[str] | Dict[str, str]
+    """Uses jinja2 to render possible template variable - kwargs will be defined global variables"""
+
+    # use recurrence for collections
+    if isinstance(tpl, list):
+        return [render_val(item, **kwargs) for item in tpl]
+    if isinstance(tpl, dict):
+        return {k: render_val(v, **kwargs) for k, v in tpl.items()}
+
+    if isinstance(tpl, str):
+        rendered = jinja2.Template(tpl, undefined=jinja2.StrictUndefined).render(env=os.environ, **kwargs)
+        if tpl!= rendered:
+            logging.debug("render_val: '%s' -> '%s'", tpl, rendered)
+        return rendered
+
+    # shortcut for lazy callers - return value with original type
+    return tpl
+
+
+def str2bool(v: str | bool) -> bool:
+    """False is denoted by empty string or any literal sensible false values"""
+    if isinstance(v, bool):
+        return v
+    return v.lower() not in ("", "no", "false", "n", "0")
+
+
+def nvm_to_dict(nvm: List[FlashMemory]) -> Dict[str, Dict[str, Any]]:
+    """Convert NVM to basic data types for jinja2 templates resolution"""
+    nvm_dict = {f.name: {p.name: p for p in f.parts} for f in nvm}
+    for flash in nvm:
+        nvm_dict[flash.name]['_meta'] = {f.name: getattr(flash, f.name) for f in fields(flash)}
+
+    return nvm_dict
+
+
+def parse_plo_script(nvm: List[FlashMemory], script_name: str) -> PloScript:
+    """Parse YAML/jinja2 plo script and return it as PloScript object"""
+
+    nvm_dict = nvm_to_dict(nvm)
+
+    with open(script_name, "r", encoding="utf-8") as f:
+        script_dict = yaml.safe_load(f)
+
+        # render templates in basic plo script attributes
+        plo_param_names = [f.name for f in fields(PloScript) if f.init]
+        plo_kwargs = {k: render_val(script_dict.get(k), nvm=nvm_dict) for k in plo_param_names if k in script_dict}
+        script = PloScript(**plo_kwargs)
+
+        tpl_context = {'nvm': nvm_dict, 'script': script}
+        for cmd in script_dict['contents']:
+            if isinstance(cmd, str):
+                cmd_rendered = render_val(cmd, **tpl_context)
+                cmddef = PloCmdFactory.build(cmd_rendered)
+            else:
+                # render all values
+                args = {k: render_val(v, **tpl_context) for k, v in cmd.items()}
+                enabled = args.pop('if', True)
+                if not str2bool(enabled):
```

Comment:
jinja2 template rendering always return `str` (if it would be invoked, `render_val` will return `bool` if `bool` would be an input, but eg. `{{ true }}` will render to `"True"` (instance of `str`)